modules/core/src/smithy4s/Endpoint.scala

```diff
@@ -66,4 +66,99 @@ object Endpoint {
 
   }
 
+  type ForOperation[Op[_, _, _, _, _]] = {
+    type e[I, E, O, SI, SO] = Endpoint[Op, I, E, O, SI, SO]
+  }
+
+  final case class Builder[I, E, O, SI, SO] private (
+      private val baseId: ShapeId,
+      private val baseHints: Hints,
+      private val baseInput: Schema[I],
+      private val baseOutput: Schema[O],
+      private val baseErrorable: Option[Errorable[E]],
+      private val baseStreamedInput: StreamingSchema[SI],
+      private val baseStreamedOutput: StreamingSchema[SO]
+  ) {
+
+    def withId(id: ShapeId): Builder[I, E, O, SI, SO] = copy(baseId = id)
+
+    def mapId(f: ShapeId => ShapeId): Builder[I, E, O, SI, SO] =
+      copy(baseId = f(baseId))
+
+    def withHints(hints: Hints): Builder[I, E, O, SI, SO] =
+      copy(baseHints = hints)
+
+    def mapHints(f: Hints => Hints): Builder[I, E, O, SI, SO] =
+      copy(baseHints = f(baseHints))
+
+    def withInput(input: Schema[I]): Builder[I, E, O, SI, SO] =
+      copy(baseInput = input)
+
+    def mapInput(f: Schema[I] => Schema[I]): Builder[I, E, O, SI, SO] =
+      copy(baseInput = f(baseInput))
+
+    def withOutput(output: Schema[O]): Builder[I, E, O, SI, SO] =
```

Comment:
**question:** should we allow using a schema of a different type (another `O`, in this case)? I'm thinking yes, because it's certainly possible.

However, when that's used in conjunction with a service builder, I think we'd be pretty limited here - not only because `PolyFunction5` forbids changing type parameters, but also I don't think we can change the operation type (which would be necessary to keep things like `def endpoint: (I, Endpoint)` working).

Reply:
I think there could eventually be two flavours of builders : 

1. One for the static usecase, that respects all the types, which probably means that the only possible transformations will be hints and labels 
2. One that would be more useful in the context of dynamic services, where we could amend both the types involved in operations, and the list of operations 

I believe this PR is only dealing with point 1 (which is totally fine). In this context, you shouldn't allow for different types.